fsql/deser.py

```diff
@@ -75,28 +128,42 @@ def detect_format(self, url: str) -> InputFormat:
             return InputFormat.from_url(url.split(".")[-1])
 
     @abstractmethod
-    def read_single(self, partition: Partition, fs: AbstractFileSystem) -> Any:  # TODO generics on the Any
+    def read_single(self, partition: Partition, fs: AbstractFileSystem) -> PartitionReadOutcome:
         raise NotImplementedError("abc")
 
     @abstractmethod
-    def concat(self, data: Iterable[Any]) -> Any:  # TODO generics on the Any
+    def concat(self, outcomes: Iterable[DataObject]) -> DataObject:
         raise NotImplementedError("abc")
 
     def read_and_concat(
         self, partitions: Iterable[Partition], fs: AbstractFileSystem
-    ) -> Any:  # TODO generics on the Any
-        with ThreadPoolExecutor(max_workers=32) as tpe:  # TODO configurable
-            data = tpe.map(partial(self.read_single, fs=fs), partitions)
-        return self.concat(data)
-
-
-class PandasReader(DataReader):
+    ) -> Union[DataObject, DataObjectRich]:
+        # TODO it is profoundly unfortunate that the return type is Union. Ideally, it would be a generic
+        # type T that is bound with this union, and determined via lazy_error parameter which would be
+        # not a boolean but instead a function [DataObject, T[DataObject]]. Alas, I was not able to pythonize
+        # that. Note this would then apply to the api's signature as well
+        with ThreadPoolExecutor(max_workers=32) as tpe:  # TODO configurable worker count
+            partition_read_outcomes = tpe.map(partial(self.read_single, fs=fs), partitions)
+            if self.lazy_errors:
+                data_objects, failures = reduce(flatten_biiterator, partition_read_outcomes)
+                # no idea why but mypy complains here:
+                # Argument 1 to "DataObjectRich" has incompatible type "DataObject"; expected "DataObject"
+                return DataObjectRich(self.concat(data_objects), failures)  # type: ignore
```

Comment:
Hmm, very interesting. When I replaced NamedTuple with dataclass (decorator) or even `collections.namedtuple(...)`, mypy was not complaining. I also found this https://bugs.python.org/issue36517 (and few other issues with NamedTuple and multiple inheritance, but this was for python 3.9 and should be already fixed). If you want to explore it more, feel free to do it. But we can look at it later, it's not some critical issue. I will merge after the weekend.